### PR TITLE
Add activation quantization support to per-channel quantized linear layers

### PR DESCRIPTION
**Pull request by lsy323**
Activation quantization is only supported with per-channel quantized model. 

Enable activation quantization with per-channel quant by using the flag `--quantize_activation=True`

The activation will be quantized to int8 and then do a int8 x int8 matmul operation. We need to call `lax.dot_general` because with torch matmul ops we cannot control the output dtype (int8 by default, and the output is easy to overflow). We use int32 as accumulation dtype to avoid overflow.

The correctness is verified in unit tests and llama/gemma model. Now get same performance on 7B int8 per-channel BS=32. In depth investigation is needed to understand the performance impact.

### CODE REVIEW

File: jetstream_pt/quantize.py [hunk 1]
```diff
@@ -95,3 +97,63 @@ def load_q_weight_helper(w_q, scale, zp=None, block_size=-1):
...
+  out = jax.lax.dot_general(
+      out, weight_scaler, dimension_numbers=([(0), (0)], [(2), (1)])
+  )
+  out = jax.lax.transpose(out, [1, 0])
+  out = out.reshape((bs, -1) + out.shape[1:])
+  return out
+
+
+def blockwise_jax_kernel_einsum_flatten(
+    inputs, weight, weight_scaler, zero_point
```
Comment (FanhaiLu1) [inline jetstream_pt/quantize.py:143]: Do you handle zero_point is not None case?

File: jetstream_pt/layers.py [hunk 2]
```diff
@@ -149,32 +172,37 @@ def __init__(
...
-      block_size=128,
-      n_bit=8,
+      quant_config=QuantizationConfig(),
   ):
     super().__init__()
     self.in_features = in_features
     self.out_features = out_features
 
     # Use dot general instead of einsum
     # Use dot general is slow now.
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/layers.py:182]: This should be an XLA issue I think, using `dot_general` and `einsum` should have the same semantics

File: jetstream_pt/layers.py [hunk 3]
```diff
@@ -115,23 +124,37 @@ def quantize_weight_from_nn_linear(self, weight):
...
     if not self.run_fake_quantize:
-      if self.is_symmetric:
-        return torch.mul(F.linear(inputs, self.weight), self.weight_scaler)
+      if self.quantize_activation:
+        inputs, act_s, _ = quantize_tensor(inputs, reduce_axis=(2,))
+      if not self.quantize_activation:
+        result = F.linear(inputs, self.weight)
       else:
-        out = torch.mul(F.linear(inputs, self.weight), self.weight_scaler)
+        result = torchjax.call_jax(
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/layers.py:139]: Here we have to call jax because we need to do `dot(int8, int8)->int32`. This semantic cannot be represented in torch now. In torch, the inferred output dtype of 2 `int8` operands will be `int8`, causing overflow on the dot result. The `dot_general` in JAX support specifying output dtype, hence we use it here.

Let me add a comment here.

File: jetstream_pt/layers.py [hunk 4]
```diff
@@ -211,112 +243,37 @@ def quantize_weight_from_nn_linear(self, weight):
...
-      else:
-        return torchjax.call_jax(
-            WeightOnlyBlockwiseQuantizedLinear.blockwise_jax_kernel,
-            inputs,
-            self.weight,
-            self.weight_scaler,
-            self.zero_point,
-        )
+        ), "Blockwise quantized linear doesn't support zero_point in dot_general or einsum flattened implementation."
+      blockwise_matmul_kernel = (
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/layers.py:257]: Thanks, this will be cleaner, let me update

File: jetstream_pt/quantize.py [hunk 5]
```diff
@@ -95,3 +97,63 @@ def load_q_weight_helper(w_q, scale, zp=None, block_size=-1):
...
+  inputs = inputs.reshape(inputs_new_shape)
+  out = jnp.einsum("scz,bdsc->bdsz", weight, inputs)
+  out = jnp.einsum("bdsz,sz->bdz", out, weight_scaler)
+  if zero_point is not None:
+    zp_out = jnp.einsum("bdsc,sz->bdz", inputs, zero_point)
+    out = out - zp_out
+  return out
+
+
+def blockwise_jax_kernel_dot_general(inputs, weight, weight_scaler, zero_point):
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/quantize.py:120]: No, we don't need to call jax for it. Thanks for the heads up. Since I'm moving the existing kernel implementation to this new file, I will switch to torch in the following PR.

File: jetstream_pt/layers.py [hunk 6]
```diff
@@ -115,23 +124,37 @@ def quantize_weight_from_nn_linear(self, weight):
...
+        weight, (1,), self.n_bit, self.is_symmetric_weight, block_size=-1
     )
     w_dq = dequantize_tensor(w_q, scale, zp)
     self._load_quantized_weights(w_q, scale, zp)
 
   def forward(self, inputs):
     if not self.run_fake_quantize:
-      if self.is_symmetric:
-        return torch.mul(F.linear(inputs, self.weight), self.weight_scaler)
+      if self.quantize_activation:
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/layers.py:134]: No, we cannot move this code to else. This is an extra step for activation quant

File: jetstream_pt/quantize.py (hunk 1 above)
Comment (lsy323) [reply jetstream_pt/quantize.py:143]: Yes, there is an assertion on the caller side https://github.com/google/jetstream-pytorch/blob/main/jetstream_pt/layers.py#L297-L299

File: jetstream_pt/quantize.py [hunk 7]
```diff
@@ -95,3 +97,63 @@ def load_q_weight_helper(w_q, scale, zp=None, block_size=-1):
...
+  weight = weight.astype(jnp.int8)
+  block_size = weight.shape[1]
+  inputs_shape = inputs.shape
+  inputs_new_shape = inputs_shape[:-1] + (
+      inputs_shape[-1] // block_size,
+      block_size,
+  )
+  inputs = inputs.reshape(inputs_new_shape)
+  out = jnp.einsum("scz,bdsc->bdsz", weight, inputs)
+  out = jnp.einsum("bdsz,sz->bdz", out, weight_scaler)
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/quantize.py:113]: Because scaler has shape `[IN_DIM / BLOCK_SIZE, OUT_DIM]`. This is because with blockwise quant, every `BLOCK_SIZE` elements share a bf16 scaler.

Other Matmul inputs: `x: [BS, SEQ, IN_DIM]`, `w: [IN_DIM / BLOCK_SIZE, BLOCK_SIZE, OUT_DIM]` (`w` has shape `[IN_DIM, OUT_DIM]` for bf16 weight)

The reduction operation is split into 2 einsum since the reduction axis is split into 2.


With the above inputs, the matmul becomes:

1. Reshape `x` to `[BS, SEQ, IN_DIM / BLOCK_SIZE, BLOCK_SIZE]`. The contract dim is split into `IN_DIM / BLOCK_SIZE` and `BLOCK_SIZE`.
2. Reduce along the `BLOCK_SIZE` dim: `x: [BS, SEQ, IN_DIM / BLOCK_SIZE, BLOCK_SIZE] @ w: [IN_DIM / BLOCK_SIZE, BLOCK_SIZE, OUT_DIM] => matmul_out [BS, SEQ, IN_DIM / BLOCK_SIZE, OUT_DIM]`
3. Reduce along `IN_DIM / BLOCK_SIZE` dim `matmul_out [BS, SEQ, IN_DIM / BLOCK_SIZE, OUT_DIM] @ scaler: [IN_DIM / BLOCK_SIZE, OUT_DIM] => matmul_out [BS, SEQ, OUT_DIM]`

**Review by FanhaiLu1 — APPROVED** [review]

File: jetstream_pt/quantize.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (qihqi) [reply jetstream_pt/quantize.py:113]: thanks, this is very clear.

**Review by qihqi — APPROVED** [review]
